### PR TITLE
SNOW-1000284: Add schema support for structure types

### PR DESCRIPTION
**Pull request by sfc-gh-jrose**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-1000284

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [x] I am adding a new dependency

3. Please describe how your code solves the related issue.

   This PR adds relevant changes needed to support Structured Types. The goal of this commit is to enable reading structured data from iceberg tables. Writing new tables and support for non-iceberg tables is a future goal of the project.

The change does the following:

1. metadata that contains fields for relevant types (MAP, OBJECT, ARRAY) signals that structured types are enabled and being returned.
2. When receiving that metadata the relevant types are marked as structured types using a new new parameter. Users that would like to make new dataframes with structured types will similarly need to set the structured_types parameter on the schema objects. In the future we will enable this behavior by default.
3. When marked as structured the relevant types will generate slightly different queries for sprocs and udfs which allows them to specify data is structured for both input and output.

I've added test cases that test the following:

1. Returned data for iceberg and non-iceberg data can be handled with the new metadata. We don't expect non-iceberg structured data yet, but it couldn't hurt to be prepared for it.
2. Sprocs can take structured data as input and provide it as output. The same goes for udfs.
3. Subfields in structured objects/maps and indexes in structured arrays can be accessed with regular dataframe semantics.
4. We're able to return structured data from requests for pandas dataframes from iceberg tables.

Merge Notes:
This change cannot merge until the next python-connector version is released. I will need to add it as a dependency to support the new pandas changes.

This change also will not pass merge gates until I figure out how to add permissions to the test runner in order to access the external volume that we're using for iceberg tests. Maybe @sfc-gh-yixie can help me figure that out.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/type_utils.py [hunk 1]
```diff
@@ -144,6 +175,8 @@ def convert_sf_to_sp_type(
         return VariantType()
     if column_type_name == "OBJECT":
         return MapType(StringType(), StringType())
+    if column_type_name == "MAP":
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/type_utils.py:178]:
```suggestion
    if column_type_name in ["MAP", "OBJECT"]:
```

File: src/snowflake/snowpark/_internal/udf_utils.py [hunk 2]
```diff
@@ -1163,7 +1163,7 @@ def create_python_udf_or_sp(
 
     if replace and if_not_exists:
         raise ValueError("options replace and if_not_exists are incompatible")
-    if isinstance(return_type, StructType):
+    if isinstance(return_type, StructType) and return_type.structured is False:
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/udf_utils.py:1166]:
```suggestion
    if isinstance(return_type, StructType) and not return_type.structured:
```

File: tests/integ/scala/test_datatype_suite.py [hunk 3]
```diff
@@ -229,6 +294,170 @@ def test_dtypes(session):
     ]
 
 
+@pytest.mark.parametrize(
```
Comment (sfc-gh-jdu) [inline tests/integ/scala/test_datatype_suite.py:297]: should we also skip these tests in stored proc? Does stored proc connector have the corresponding change for struct type?

File: tests/utils.py [hunk 4]
```diff
@@ -76,6 +76,12 @@
 IS_NOT_ON_GITHUB = os.getenv("GITHUB_ACTIONS") != "true"
 # this env variable is set in regression test
 IS_IN_STORED_PROC_LOCALFS = IS_IN_STORED_PROC and os.getenv("IS_LOCAL_FS")
+STRUCTURED_TYPE_ENVIRONMENTS = {"dev", "aws"}
+IS_STRUCTURED_TYPES_SUPPORTED = (
+    os.getenv("cloud_provider", "dev") in STRUCTURED_TYPE_ENVIRONMENTS
```
Comment (sfc-gh-jdu) [inline tests/utils.py:81]: cloud_provider is set on github, right? We also have some tests running on jenkins, so they will just be skipped on jenkins?

File: tests/integ/scala/test_datatype_suite.py [hunk 5]
```diff
@@ -33,7 +36,69 @@
...
+                    nullable=True,
+                ),
+                StructField(
+                    "ARR", ArrayType(DoubleType(), structured=True), nullable=True
+                ),
+            ]
+        ),
+        id="structured-types-enabled",
+    ),
+    False: pytest.param(
```
Comment (sfc-gh-jdu) [inline tests/integ/scala/test_datatype_suite.py:80]: we finally will remove this example once all deployments supporting struct types?

Comment (sfc-gh-jrose) [reply tests/integ/scala/test_datatype_suite.py:80]: Yes, at some point in the future we will make a BCR that turns structured types on by default. That will probably be the best time to remove this.

File: tests/integ/scala/test_datatype_suite.py (hunk 3 above)
Comment (sfc-gh-jrose) [reply tests/integ/scala/test_datatype_suite.py:297]: By the time I merge this change it should have the corresponding change.

File: tests/utils.py (hunk 4 above)
Comment (sfc-gh-jrose) [reply tests/utils.py:81]: I'm not sure how that gets set. I know that in tox.ini it's passed in regardless of environment, but that doesn't mean it is set.

Comment (sfc-gh-jdu) [reply tests/utils.py:81]: so we will skip these tests on jenkins right now? e.g., https://ci-dev-142.int.snowflakecomputing.com/job/SnowparkPythonClientRegressRunner/

Comment (sfc-gh-jrose) [reply tests/utils.py:81]: Yes. Once the parameter for enabling structured types is on by default we will need to make sure these tests work in all environments, but for now sfctest0 is the only environment that has them enabled. It is also the only environment that we have iceberg test infrastructure set up.

Comment (sfc-gh-jdu) [reply tests/utils.py:81]: got it, thanks. Then maybe create a jira and add a TODO here to make sure these conditions will be finally removed?

**Review by sfc-gh-jdu — APPROVED** [review]

File: tests/integ/scala/test_datatype_suite.py [hunk 6]
```diff
@@ -33,7 +36,69 @@
...
+        [
+            ("MAP", "map<string(16777216),bigint>"),
+            ("OBJ", "struct<string(16777216),double>"),
+            ("ARR", "array<double>"),
+        ],
+        StructType(
+            [
+                StructField(
+                    "MAP",
+                    MapType(StringType(16777216), LongType(), structured=True),
```
Comment (sfc-gh-yixie) [inline tests/integ/scala/test_datatype_suite.py:59]: Do we have to explicit use the length 16777216? This number may change soon.

**Pull request merged**